### PR TITLE
Add ProjectType

### PR DESCRIPTION
**Pull request by knolleary**
Implements the core of #380 

The high-level description is provided here: https://github.com/flowforge/flowforge/issues/380#issuecomment-1106426912

 - Adds `/api/v1/project-types/` end points for managing the types
 - Adds `projectType` as a field of stacks - stacks are owned by a project type.
 - When billing enabled, the ProjectType has ` billingProductId` and `billingPriceId` properties. When a project is added/removed from a subscription, we use those properties if they exist - otherwise default to the values provided in the yml file.

Will update with more notes later.

UX demo (with some bugs...): https://www.youtube.com/watch?v=vML-U0sgqik

## Remaining Work

*Prior to merging this PR*

 - ~Delete ProjectType api end point + UI~
 - ~'Change Stack' workflow must restrict the choice of stacks to those of the project type.~
 - More testing of legacy projects

*Can be considered for future iterations*

 - #595

### CODE REVIEW

**Comment by sammachin** [comment]
Are the stripe ID editable after a type has been created? I'm thinking through what that could mean and if it would break something should they get changed

**Comment by knolleary** [comment]
> Are the stripe ID editable after a type has been created? I'm thinking through what that could mean and if it would break something should they get changed

No - properties cannot be edited if there are any projects of this type.

If the ProjectType has been created, but no projects yet created using it, then they are editable.

File: forge/db/migrations/20220623-01-add-project-type.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+            onUpdate: 'cascade'
+        })
+        await context.addColumn('ProjectStacks', 'ProjectTypeId', {
+            type: DataTypes.INTEGER,
+            references: { model: 'ProjectTypes', key: 'id' },
+            onDelete: 'set null',
+            onUpdate: 'cascade'
+        })
+    },
+    down: async (context) => {
```
Comment (ZJvandeWeg) [inline forge/db/migrations/20220623-01-add-project-type.js:41]: I would recommend writing a down migration in case a deploy goes sideways.

**Comment by knolleary** [comment]
For a legacy project, when they go to the Danger section of Settings, they are shown a one-time option to set the project type - and the option to change stack is disabled.

image

The option shows them the list of available project types for them to pick from:

image

Once the project has a project-type set, we hide this whole option. Changing Project Type is part of #595 and not in scope of today's work.

**Comment by knolleary** [comment]
I've looked at a few options around the migration. Getting the balance between work for the user, work for the admin and work we can do in the background magically.

What I describe above is the user-drive approach. Whilst writing out the description, I have identified an admin-driven way to set the project-types so the user doesn't have to:

After upgrading to 0.7, the platform admin will have to create a Project Type in the system. They then have to assign the *existing* stacks to that project type. When they do that second step, we can safely update each project using a given stack to match. As long as the billing details match - which the admin needs to get right - then nothing will be impacted. With that approach, the user will never know about the change.

**Comment by knolleary** [comment]
I'm 99% happy with this.

A couple more scenarios that need testing against before we green light this.

 - duplicating a project
 - export to project

**Review by hardillb — COMMENTED** [review]
Apart from the delete Project Type not working, all looks good.

File: frontend/src/pages/admin/ProjectTypes/index.vue [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+            }
+        },
+        showCreateProjectTypeDialog () {
+            this.$refs.adminProjectTypeEditDialog.show()
+        },
+        showEditProjectTypeDialog (projectType) {
+            this.$refs.adminProjectTypeEditDialog.show(projectType)
+        },
+        showConfirmProjectTypeDeleteDialog (projectType) {
+            // this.$refs.adminStackDeleteDialog.show(stack)
```
[the remark this answers is not in the document]
Comment (knolleary) [reply frontend/src/pages/admin/ProjectTypes/index.vue]: Ooops - having got the api and tests written, I didn't come back and wire up the UI. Now done.

File: forge/db/test-data.js [hunk 3]
```diff
@@ -1,30 +1,41 @@
...
-        const team3 = await app.db.models.Team.create({ name: 'CTeam' })
-
-        await team1.addUser(userAlice, { through: { role: Roles.Owner } })
-        await team1.addUser(userBob, { through: { role: Roles.Member } })
-        // await team1.addUser(userChris, { through: { role:Roles.Member } });
-
-        await team2.addUser(userBob, { through: { role: Roles.Owner } })
-        await team2.addUser(userAlice, { through: { role: Roles.Member } })
-
-        await team3.addUser(userAlice, { through: { role: Roles.Owner } })
```
[the remark this answers is not in the document]
Comment (hardillb) [reply forge/db/test-data.js:27]: Yep, with this commented out, it doesn't actually do anything anymore

File: frontend/src/pages/admin/ProjectTypes/index.vue (hunk 2 above)
[the remark this answers is not in the document]
Comment (hardillb) [reply frontend/src/pages/admin/ProjectTypes/index.vue]: Looks good

**Pull request merged**